contrib/database/sql/sql.go

```diff
@@ -12,49 +12,169 @@
 package sql
 
 import (
+	"context"
 	"database/sql"
 	"database/sql/driver"
 	"errors"
+	"math"
+	"reflect"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql/internal"
 )
 
+// registeredDrivers holds a registry of all drivers registered via the sqltrace package.
+var registeredDrivers = &driverRegistry{
+	keys:    make(map[reflect.Type]string),
+	drivers: make(map[string]driver.Driver),
+	configs: make(map[string]*config),
+}
+
+type driverRegistry struct {
+	// keys maps driver types to their registered names.
+	keys map[reflect.Type]string
+	// drivers maps keys to their registered driver.
+	drivers map[string]driver.Driver
+	// configs maps keys to their registered configuration.
+	configs map[string]*config
+}
+
+// registered checks if the named driver has already been registered.
+func (d *driverRegistry) registered(name string) bool {
+	_, ok := d.configs[name]
+	return ok
+}
+
+// add adds the driver and config to the driver registy.
+func (d *driverRegistry) add(name string, driver driver.Driver, cfg *config) {
+	if d.registered(name) {
+		return
+	}
+	d.keys[reflect.TypeOf(driver)] = name
+	d.drivers[name] = driver
+	d.configs[name] = cfg
+}
+
+// name returns the name of the driver stored in the registry.
+func (d *driverRegistry) name(driver driver.Driver) (string, bool) {
```

Comment:
From what I see this isn't really used except to get a name and then pass it straight to `(*driverRegistry).config`, so maybe we can simplify by removing this method and changing `config` to be:
```go
func (d *driverRegistry) config(driver driver.Driver) (cfg *config, ok bool) {
    name, ok := d.keys[reflect.TypeOf(driver)]
    if !ok {
        return false
    }
    cfg, ok = d.configs[name]
    return cfg, ok
}
```
What do you think?

Reply:
The name gets passed down to `traceParams` and eventually in the span name.
It could be `nameAndConfig(driver.Driver) (string, *config, bool)`.

I think it's simple enough for now.

Reply:
Oh, right. Ok, I see. I guess name could be part of config, but that'd make things a bit confusing and inconsistent. Let's leave it then. Resolving.